### PR TITLE
[liburing] create a new port

### PR DESCRIPTION
**Pull request by luncliff**
### What does your PR fix?  

There was no port request for this library. But I think some Linux people are interested in it.

https://github.com/axboe/liburing (It requires kernel version 5.1+.)

### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  

Targets Linux triplets.

* `x64-linux`

Wish other contributors who use Linux with ARM processor can check this works for those triplets.  
[Seems like `CC`/`CXX`/`BUILD_ARCH` customization is required](https://github.com/axboe/liburing/blob/master/.travis.yml).

* `arm-linux`
* `arm64-linux`

### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Checked. 

I saw Docker images target Linux Kernel 4.x in late 2019.  
Considering to check the kernel version in the portfile and warn it.

### CODE REVIEW

**Comment by NancyLi1013** [comment]
Is there anything else that needs to be done for this PR @luncliff?

**Comment by luncliff** [comment]
Sorry for the pause, Nancy.
I asked my friend for trying this, but not getting any feedback. I checked
symbols with `nm` but not sure the port works correctly.
Let me check some uring samples in GitHub in the upcoming weekend.

**Comment by luncliff** [comment]
I tested build/link of this port with both `VCPKG_LIBRARY_LINKAGE` static/dynamic.
Should be like below.

```console
user@host:build-dbg$ ldd ./test_suite 
        linux-vdso.so.1 (0x00007fff1a14f000)
        liburing.so.2 => /.../vcpkg/installed/x64-linux/debug/lib/pkgconfig/../../lib/liburing.so.2 (0x00007ffbe4c1e000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007ffbe4be5000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007ffbe4a04000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007ffbe48b5000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007ffbe489a000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ffbe46a8000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ffbe4d7e000)
```

Plus, I added `usage` for CMake [`FindPkgConfig`](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html) module.
There was no existing references for `FindPkgConfig` integration so I wrote it.
Any suggestion for it will be welcomed. :D

File: ports/liburing/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,41 @@
+vcpkg_fail_port_install(MESSAGE "'${PORT}' is for Linux" ON_TARGET "windows" "uwp" "osx" "ios" "android")
```
Comment (NancyLi1013) [inline ports/liburing/portfile.cmake:1]:
```suggestion
vcpkg_fail_port_install(ON_TARGET "windows" "uwp" "osx" "ios" "android")
```
This message is not necessary.

File: ports/liburing/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    )
+else()
+    file(REMOVE ${CURRENT_PACKAGES_DIR}/debug/lib/liburing.so
+                ${CURRENT_PACKAGES_DIR}/debug/lib/liburing.so.2
+                ${CURRENT_PACKAGES_DIR}/debug/lib/liburing.so.2.0.0
+                ${CURRENT_PACKAGES_DIR}/lib/liburing.so
+                ${CURRENT_PACKAGES_DIR}/lib/liburing.so.2
+                ${CURRENT_PACKAGES_DIR}/lib/liburing.so.2.0.0
+    )
+endif()
```
Comment (NancyLi1013) [inline ports/liburing/portfile.cmake:40]: Can we only build one type of library? static or dynamic? Instead of removing the binaries here.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/liburing/portfile.cmake (hunk 2 above)
Comment (luncliff) [reply ports/liburing/portfile.cmake]: I checked the `Makefile`s and it seems like `liburing.so` requires `liburing.a` for it. Would you review the changes of 8a0e6e3?

File: ports/liburing/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    SOURCE_PATH ${SOURCE_PATH}
+    COPY_SOURCE
+    NO_DEBUG
+)
+# note: {SOURCE_PATH}/src/Makefile makes liburing.so from liburing.a.
+#   For static, disable .so using the environment variable ENABLE_SHARED.
+#   For dynamic, remove intermediate file liburing.a when install is finished.
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+    set(ENV{ENABLE_SHARED} "0")
+endif()
```
Comment (JackBoosY) [inline ports/liburing/portfile.cmake:25]: This should be handled in `vcpkg_configure_make.cmake`:
https://github.com/microsoft/vcpkg/blob/7c55ecac266fc3e554bc315053dc45c11ec811af/scripts/cmake/vcpkg_configure_make.cmake#L496-L500

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/liburing/portfile.cmake (hunk 3 above)
Comment (luncliff) [reply ports/liburing/portfile.cmake:25]: Got it. I will make a patch for this.

**Review by JackBoosY — APPROVED** [review]
Did you test the usage?

**Comment by luncliff** [comment]
@JackBoosY Yeap. with https://github.com/microsoft/vcpkg/pull/17623#issuecomment-842001199

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM now, thanks for your adding this new port @luncliff.

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

File: ports/liburing/usage [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+    include(FindPkgConfig)
+    pkg_check_modules(URING liburing>=2.0)
+
+    target_include_directories(main PRIVATE ${URING_INCLUDE_DIRS})
+    target_link_libraries(main PRIVATE ${URING_LINK_LIBRARIES})
+
+If you are using CMake v3.13 or higher:
+
+    target_link_directories(main PRIVATE ${URING_LIBRARY_DIRS})
```
Comment (strega-nil-ms) [inline ports/liburing/usage:11]:
```suggestion
    include(FindPkgConfig)
    pkg_check_modules(liburing REQUIRED IMPORTED_TARGET GLOBAL liburing>=2.0)
    
    target_include_directories(main PRIVATE PkgConfig::liburing)
```

**Comment by strega-nil-ms** [comment]
@luncliff could you make the versions changes? I apparently can't push to your branch...

**Comment by luncliff** [comment]
> @luncliff could you make the versions changes? I apparently can't push to your branch...

Of course! Thanks for the check and commit. :) 
I will update the part and check import steps again.

**Comment by luncliff** [comment]
@strega-nil-ms please ping me when it's reviewed. :)

**Review by NancyLi1013 — APPROVED** [review]

**Review by strega-nil — APPROVED** [review]

**Pull request merged**